[core] Support for generated columns

CaselIT opened the pull request:
Support for generated column

this it just the core parts of #4896
part of #4894

CaselIT committed:
Initial implementation of a generated column

Add a Generated schema item that can be used in a column definition
Add default compiler implementation

ref https://github.com/sqlalchemy/sqlalchemy/issues/4894

Fixes: #<issue number>

CaselIT committed:
rename generated to computed

CaselIT committed:
fix: generated precedes null | not null

CaselIT committed:
feat: add dialect support

CaselIT committed:
fix mssql: do not render NULL when a computed column is nullable

CaselIT committed:
text: add test to computed columns

CaselIT committed:
update test style

CaselIT committed:
test: avoid using pytest directly

CaselIT commented:
if https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1523 lands before this is done, I think we could refactor the test using `combinations`

zzzeek commented:
OK I've merged `@testing.combinations` which ultimately calls pytest.parametrize.  want to give that a shot?  also do we want to close #4896 ?

CaselIT commented:
> want to give that a shot?

Nice, I'll convert the test later then

> also do we want to close #4896 ?

I'm using that pr to implement the orm part of the generated column. It continues after this pr.
I've done a first rough implementation, but I'm not sure if that is the best way to do it.

I can open a different pr if you prefer.

CaselIT committed:
Merge branch 'master' into generated_column_core

CaselIT commented:
I'm looking at combinations. What does `a` stands for?

CaselIT committed:
use combinations in the tests

CaselIT committed:
add missing option to documentation of combinations

CaselIT commented:
I've updated the tests.
I think you can review

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work to try to get revision 19a1b032683a6ee600df9e1f65ef0d3cec88cbeb of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
New Gerrit review created for change 19a1b032683a6ee600df9e1f65ef0d3cec88cbeb: https://gerrit.sqlalchemy.org/#/c/sqlalchemy/sqlalchemy/+/1539

zzzeek commented:
this is looking great.  the next thing that should be within this commit are backend round trip tests in lib/sqlalchemy/testing/suite/ ....maybe test_select.py for now.   it would be a new ComputedColumnTest and would do a create / drop and also round trip a row from a SELECT to see that it gets something back, this is more to test that the DDL actually does the right thing on the backend and that there is no driver weirdness when fetching it or anything like that.

I can do it if you'd like if you want to be done, or you can keep on going with it, thanks for your efforts!

CaselIT commented:
Nice, thanks for the review.

I'll check pep8 and why py2.7 is not working. (Also do you plan to drop support for it in the near future?)

I'll try adding the select test later today.

zzzeek commented:
the python 2 plan is that "SQLAlchemy 2.0" will not support Python 2.

SQLAlchemy 2 is tentatively planned for the major version after 1.4 (note SQLAlchemy "major versions" have up til now been the "minor version" number via semver), however 1.4 is getting quite full with a lot of big changes that precede the 2.0 changes, so there might even be a 1.5.     I can see *maybe* dropping py2k for 1.5 if it goes that way.   

anyway, we still have at least a year to go.

zzzeek commented:
for the pep8 stuff I use pre_commit with black / zimports / flake8, if you look in .pre-commit-config.yaml.   that will do everything except for formatting issues in docstrings.  the flake8 check, with plugins instlaled, will show if there's anything that wont pass the extended flake8 checks in the commit.

CaselIT commented:
> the python 2 plan is ..

seems clear

ok regarding pep8

CaselIT committed:
fix python 2.7 syntax

CaselIT committed:
add backend test for combuted columns

zzzeek commented on test/dialect/mssql/test_compiler.py:
```diff
@@ -1195,15 +1195,15 @@ def test_try_cast(self):
 
     @testing.combinations(
-        ("no_persisted", "", ...),
+        ("no_persisted", "", Ellipsis),
```
I had never heard of this object before.  what does it mean in this context?

CaselIT replied:
Is is used sometimes in the indexes, like numpy uses it sometimes. It is not very used otherwise.

Here it is just a flag that I use to check it `persisted` should be added to the attributes.
I guess `skip_arg` or a `flag=object()` would also work here

CaselIT replied:
I guess it has also been repurposed a bit in the typing library to generate empty functions/classes and specify some types

sqla-tester replied:
**mike bayer** (zzzeek) wrote:

python 2 is not accepting this.

CaselIT commented:
I've fixed python 2.7 (I think). It does not like `...` for `Ellipsis`. The pep8 stuff seems only regarding the import order.

I've added a couple of tests in test_select
I'm not sure if a test without sqlalchemy should be also done. I mean with the raw connection, to verify that it is not sqlalchemy that generates the columns

I've thought at what other test could be done:

- insert is covered by the select one.
- update maybe could be tested, but we are checking that the db is working correctly here, so not sure useful

zzzeek commented:
absolutely, I haven't had time to look at it.   had some time to push through small things yesterday but I'm really having a shortage of time, you are on the list

CaselIT commented:
Sorry, I did not mean to pressure you.

zzzeek commented:
no worries you've done a great job on this pr

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

recheck

sqla-tester commented on lib/sqlalchemy/dialects/mysql/base.py:
```diff
@@ -1503,6 +1503,10 @@ def get_column_specification(self, column, **kw):
 
+        if column.computed is not None:
+            # the default visit_computed_column is compatible
+            colspec.append(self.process(column.computed).strip())
```
**mike bayer** (zzzeek) wrote:

why do we need the strip() here?

CaselIT replied:
The computed compiler leaves a space at the end. This dialect is a bit different from the other because it uses a list then joins it a space at the end.

I'm not sure if the additional space can be left or if it will cause problems.

sqla-tester commented on lib/sqlalchemy/dialects/oracle/base.py:
```diff
@@ -1187,6 +1187,20 @@ def post_create_table(self, table):
+        # explicitly check for True|False since None means server default
+        if generated.persisted is True:
+            raise exc.CompileError(
+                "oracle does not support persited (stored) computed "
```
**mike bayer** (zzzeek) wrote:

typo 'persited'

sqla-tester commented on lib/sqlalchemy/dialects/sqlite/base.py:
```diff
@@ -1032,6 +1032,9 @@ def visit_empty_set_expr(self, element_types):
 
 class SQLiteDDLCompiler(compiler.DDLCompiler):
     def get_column_specification(self, column, **kwargs):
+        if column.computed is not None:
```
**mike bayer** (zzzeek) wrote:

maybe just let it get passed through to SQLite and let it raise?   because SQLite might add this feature.

CaselIT replied:
I raise the error so it is consistent with the other implementations that raise an error in case of an unsupported option. Like in oracle it raises an error if persisted is true.
If we do not raise an error here I think we also should not raise an error in case of an unsupported option in the other dialects

sqla-tester replied:
**mike bayer** (zzzeek) wrote:

well when we see an option for which we don't know what SQL to generate, that's harder to ignore because we can't have it emit an error from the DB.   it's fine to leave this as is however.

CaselIT replied:
Ok

sqla-tester commented on test/sql/test_computed.py:
```diff
@@ -0,0 +1,54 @@
+    __dialect__ = "default"
+
+    @combinations(
+        ("no_persisted", "", Ellipsis),
```
**mike bayer** (zzzeek) wrote:

we're still doing Python 2 so this has to be something that works in py2k also.

CaselIT replied:
I'll change this.
The logic of the test is that we have 4 cases:
- Persisted=true
- Persisted=false
- Persisted=None
- Persisted not provided

I use Ellipsis for the last case, to ignore the value passed to the function, and call only `Computed`

sqla-tester reviewed (commented):
**mike bayer** (zzzeek) wrote:

(5 comments)

going to rerun pep8 in a moment...some changes so far, mostly we have to figure out the "..." thing.   i haven't looked to see what this does yet can that just be None?

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

pep8-recheck

zzzeek commented:
oh crap did i not pull the most recent version in?

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work to try to get revision 9348b2d6396dea0b56c5750ab042601b810f0be0 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset 9348b2d6396dea0b56c5750ab042601b810f0be0 added to existing Gerrit review https://gerrit.sqlalchemy.org/#/c/sqlalchemy/sqlalchemy/+/1539

zzzeek commented:
ok ignore the ellipsis thing , sorry

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

Code-Review-1

the pep8 failures are import order stuff, use the zimports tool to clean them up

./test/dialect/oracle/test_compiler.py:5:1: I100 Import statements are in the wrong order. 'from sqlalchemy import Computed' should be before 'from sqlalchemy import except_'
from sqlalchemy import Computed
^
./test/dialect/oracle/test_compiler.py:33:1: I100 Import statements are in the wrong order. 'from sqlalchemy.testing import assert_raises_message' should be before 'from sqlalchemy.testing import AssertsCompiledSQL'
from sqlalchemy.testing import assert_raises_message
^
./test/sql/test_computed.py:5:1: I101 Imported names are in the wrong order. Should be assert_raises_message, AssertsCompiledSQL, combinations, fixtures
from sqlalchemy.testing import (
^

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

database failures are currently postgresql:

psycopg2.errors.SyntaxError: syntax error at or near "("
LINE 5:  area INTEGER GENERATED ALWAYS AS (side * side) STORED,

CaselIT commented:
> ok ignore the ellipsis thing , sorry

No it makes sense clarifing the test. I'll change it

> database failures are currently postgresql:

Strange, i've run the test with postgress locally. Which version of postgress is it testing against? It must be the 12+

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

(1 comment)

> CaselIT@github wrote:
 > 
 > > ok ignore the ellipsis thing , sorry
 > 
 > No it makes sense clarifing the test. I'll change it
 > 
 > > database failures are currently postgresql:
 > 
 > Strange, i've run the test with postgress locally. Which version of
 > postgress is it testing against? It must be the 12+

the tests run against PG 9.6, 10, 11 and 12.   So only the 12 is passing, so you need to have a @testing.requires.computed_columns requirement added to test/requirements.py.    Also the computed column thing should be something that 3rd party dialects can support also so we should have some tests in lib/sqlalchemy/testing/suite/ as well as a closed @testing.requires.computed_columns in lib/sqlalchemy/testing/requirements.py

CaselIT commented:
Thanks for the review.

I'll have a look on it and add them.

CaselIT committed:
fix typo

CaselIT committed:
clarify the test by avoiding using ellipsis

CaselIT committed:
fix pep8 errors

CaselIT committed:
skip computed columns on postgre < 12

CaselIT committed:
fix requirements in sqlalchemy suite

CaselIT commented:
I've fixed the typo and replaced the use of ellipsis in the test, with a cleared string 'ignore'. Also the pep8 stuff should be fixed

I've tried to add the requirements as you suggested, and it should work now event on previous postgres versions

zzzeek commented on lib/sqlalchemy/testing/suite/test_select.py:
```diff
@@ -858,3 +859,46 @@ def test_contains_autoescape_escape(self):
+
+class ComputedColumnTest(fixtures.TablesTest):
+    __backend__ = True
+    __unsupported_on__ = ("sqlite", "postgresql < 12")
```
this should say:

    __requires__ = ('computed_columns', )

zzzeek commented:
> I've fixed the typo and replaced the use of ellipsis in the test, with a cleared string 'ignore'. Also the pep8 stuff should be fixed
> 
> I've tried to add the requirements as you suggested, and it should work now event on previous postgres versions

thanks!   i''ll run this through now.   most everything is here so I can take this over whenever you are burnt out on it thanks for all the efforts

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work to try to get revision d39c521d5ac6ebfb4fb5b53846451de79752e64c of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset d39c521d5ac6ebfb4fb5b53846451de79752e64c added to existing Gerrit review https://gerrit.sqlalchemy.org/#/c/sqlalchemy/sqlalchemy/+/1539

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

MySQL reports:

(1064, "You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'GENERATED ALWAYS AS (side * side), \n\tperimeter INTEGER GENERATED ALWAYS AS (4 * ' at line 4")

I'm going to guess this might be because engine="MyISAM"?   try adding mysql_engine="InnoDB"

CaselIT commented:
> thanks! i''ll run this through now. most everything is here so I can take this over whenever you are burnt out on it thanks for all the efforts

It will provably be quicker if you can finish this up. I cannot run all the tests so it's hard to cover all the cases. Thanks

Regarding the mysql error I don't know. It may also be the version, if I'm not mistaken it's not too long that it supports the generated columns

zzzeek commented:
oh it's just mysql 5.6 doesn't have computed columns.  might be time to drop 5.6 test support.

CaselIT commented:
Thanks for making the test pass

sqla-tester commented:
**Federico Caselli** (CaselIT) wrote:

Code-Review+1

Nice

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

I've added documentation as this is a major new feature and also adjusted the docs for Computed.   feel free to give it a quick proof.  there is also a backport for 1.3.    there's a lot here this is very big.

CaselIT commented:
I've given the documentation a look and it seems clear.
Thanks for finalizing this feature.

CaselIT commented:
I thought a bit more about the documentation. I don't know if we should add a note specifing that expression supported in the computed column depends on the database. Like some databases support computed columns that reference other computed columns, some don't.

zzzeek commented:
All of these DDL things have huge lists of caveats, that's one where they'll get an immedaite error raise if that's the case, so they'll figure it out :)

CaselIT commented:
Yes, that was my guess also

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

Code-Review+2 Workflow+1

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/1539 has been **merged**. Congratulations! :)

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

Code-Review+2 Workflow+1

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/1559 has been **merged**. Congratulations! :)

CaselIT commented:
Nice
